src/main/java/org/spongepowered/api/plugin/Plugin.java

```diff
@@ -24,19 +24,21 @@
  */
 package org.spongepowered.api.plugin;
 
-import static java.lang.annotation.ElementType.TYPE;
-import static java.lang.annotation.RetentionPolicy.RUNTIME;
-
+import java.lang.annotation.ElementType;
 import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
 import java.lang.annotation.Target;
+import java.util.regex.Pattern;
 
 /**
  * An annotation used to describe and mark a Sponge plugin.
  */
-@Target(TYPE)
-@Retention(RUNTIME)
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
 public @interface Plugin {
 
+    Pattern ID_PATTERN = Pattern.compile("[a-z0-9-_]+");
```

Comment:
Having read #1083, there are a number of issues with using classes as identification (not least of which is the fact that we'd need to pre-emptively strip the class out of the annotation before the class is loaded and converted it to a string anyway. But that's besides the point for this PR, since it's a separate issue and attempting to impose a different ideology here (which seems to directly conflict with a best-practice approach of making ids as unique as possible) would be more appropriate as discussion on #1083 rather than being included in these changes.

My personal opinion is that not allowing dots in plugin ids is too restrictive and actively encourages exactly the kind of behaviour which is concerning @windy1 in #1083.